File: leecode/90-子集Ⅱ.py
def subsetsWithDup(nums):
    n = len(nums)
    nums.sort()

    def backtrack(i=0, list_one=[]):
        res.append(list_one[:])
        for j in range(i, n):
            if j > i and nums[j] == nums[j-1]:
                continue
            list_one.append(nums[j])
            backtrack(j + 1, list_one)
            list_one.pop()
    res = []
    backtrack()
    return res

File: leecode/1-500/1-100/90-子集Ⅱ.cpp
class Solution {
public:
    void backtrack(vector<vector<int>>& ans, vector<int> nums, vector<int>& tmp, int first) {
        ans.emplace_back(tmp);
        for(int i = first; i < nums.size(); ++i) {
            if (i != first && nums[i] == nums[i-1])
                continue;
            tmp.emplace_back(nums[i]);
            backtrack(ans, nums, tmp, i+1);
            tmp.pop_back();
        }
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> ans;
        vector<int> tmp;
        backtrack(ans, nums, tmp, 0);
        return ans;
    }
};